Piece.cpp

```diff
@@ -6,78 +6,132 @@
 #include "Board.hpp"
 
 #include <iostream>
+#include <cstdio>
+#include <cmath>
+
 
 Piece::Piece() {
 
-  
-  rect.setSize(this->size);
-  rect.setPosition(this->position); 
-  
+  rect.setFillColor(sf::Color::Blue);  
+  rect.setSize(size);
+  // Test knight code
+  moveset.horizontal = false;
+  moveset.vertical = false;
+  moveset.diagonal = true;
+  moveset.offsets.push_back(sf::Vector2i(1,2));
+
+  collider.width = size.x;
+  collider.height = size.y;
+
+  distributePosition();
 }
 
 
 Piece::~Piece() {
-  snapRect = nullptr;
 }
 
+void Piece::consumeMoveset(Moveset moves, bool XORMode) {
+  if (!XORMode) {
+    this->moveset = this->moveset || moves;
+  }
+}
 
-void Piece::snapToGrid(sf::RenderWindow& window, Board& board) {
 
-  this->rect.setPosition(this->position);
-  
-  this->collider.top = this->position.y;
-  this->collider.left = this->position.x;
-  this->collider.width = this->size.x;
-  this->collider.height = this->size.y;
+void Piece::snapToSector(sf::Vector2i sector, Board& board) {
+  position = sf::Vector2f( board.sectors[sector.y-1][sector.x-1].left, board.sectors[sector.y-1][sector.x-1].top );
+}
 
-  
-  for ( int i = 0; i < board.sectors.size(); i++ ) {
-    for ( int r = 0; r < board.sectors[i].size(); r++ ) {
-      
-      if ( board.sectors[i][r].contains((sf::Vector2f)sf::Mouse::getPosition(window))
-	   && sf::Mouse::isButtonPressed(sf::Mouse::Button::Left) ) {
+bool Piece::validateMove(Board &board, sf::Vector2i pos){
 
-	snapRect = &board.sectors[i][r];
+	// Check infinite directional bools 
 	
+	if (moveset.horizontal) {
+	  if (sectorPosition.y == pos.y) {
+	    std::cout << "horizontal\n";
+	    return true;
+	  }
+	}
+	else if (moveset.vertical) {
+	  if (sectorPosition.x == pos.x) {
+	    std::cout << "vertical\n";
+	    return true;
+	  }
+	}
+	else if (moveset.diagonal) {
+	  if (abs(sectorPosition.x - pos.x) ==
+	      abs(sectorPosition.y - pos.y)) {
+
+	    std::cout << "diagonal\n";
+	    return true;
+	  }
+	}
+	// Check offsets for valid moves
+            for (auto m : moveset.offsets) {  
+              if ( pos == sectorPosition+m ) {
+                return true;
+              }
+            }
+  return false;
+}
+
+void Piece::onEvent(sf::Event event, Board &board) {
+  if ( beingMoved
```

Comment:
reducing the imbrication levels would help make the code more readable.

maybe return early if the conditions are not met example:
```
if(false) {
  return;
}
// actual code
```